Lib/email/utils.py

```diff
@@ -103,12 +104,126 @@ def formataddr(pair, charset='utf-8'):
     return address
 
 
+def _iter_escaped_chars(addr):
+    pos = 0
+    escape = False
+    for pos, ch in enumerate(addr):
+        if ch == '\\' and not escape:
+            escape = True
+        else:
+            if escape:
+                yield (pos, '\\' + ch)
+            else:
+                yield (pos, ch)
+            escape = False
+    if escape:
+        yield (pos, '\\')
+
+
+def _strip_quoted_realnames(addr):
+    """Strip real names between quotes."""
+    if '"' not in addr:
+        # Fast path
+        return addr
+
+    start = None
+    remove = []
+    for pos, ch in _iter_escaped_chars(addr):
+        if ch == '"':
+            if start is None:
+                start = pos
+            else:
+                remove.append((start, pos))
+                start = None
+
+    result = []
+    pos = 0
+    for start, stop in remove:
+        result.append(addr[pos:start])
+        pos = stop + 1
+    result.append(addr[pos:])
+
+    return ''.join(result)
+
+
+def getaddresses(fieldvalues, *, strict=True):
+    """Return a list of (REALNAME, EMAIL) or ('','') for each fieldvalue.
+
+    When parsing fails for a fieldvalue, a 2-tuple of ('', '') is returned in
+    its place.
 
-def getaddresses(fieldvalues):
-    """Return a list of (REALNAME, EMAIL) for each fieldvalue."""
-    all = COMMASPACE.join(str(v) for v in fieldvalues)
-    a = _AddressList(all)
-    return a.addresslist
+    If strict is True, use a strict parser which rejects malformed inputs.
+    In this case, if the resulting list of parsed addresses is greater than
+    the number of fieldvalues in the input list, a parsing error has occurred
+    and consequently a list containing a single empty 2-tuple [('', '')] is
+    returned in its place. This is done to avoid invalid output.
+
+    Malformed input: getaddresses(['alice@example.com <bob@example.com>'])
+    Invalid output: [('', 'alice@example.com'), ('', 'bob@example.com')]
+    Safe output: [('', '')]
+
+    """
+    if not strict:
+        all = COMMASPACE.join(str(v) for v in fieldvalues)
+        a = _AddressList(all)
+        return a.addresslist
+
+    fieldvalues = [str(v) for v in fieldvalues]
+    fieldvalues = _pre_parse_validation(fieldvalues)
+    addr = COMMASPACE.join(v for v in fieldvalues)
+    a = _AddressList(addr)
+    result = _post_parse_validation(a.addresslist)
+
+    # Treat output as invalid if the number of addresses is not equal to the
+    # expected number of addresses.
+    n = 0
+    for v in fieldvalues:
+        # When a comma is used in the Real Name part it is not a deliminator.
+        # So strip those out before counting the commas.
+        v = _strip_quoted_realnames(v)
+        # Expected number of addresses: 1 + number of commas
+        n += 1 + v.count(',')
```

Comment:
Can there be escaped commas?

Reply:
The parser doesn't support escaped commas:

```py
>>> email.utils.getaddresses([r'Real\, name <alice@example.org>'])
[('', 'Real\\'), ('name', 'alice@example.org')]
```